### PR TITLE
Add ExternalEventsComponent & MetricsComponent

### PR DESCRIPTION
**Pull request by caisq**
- listens for external keypresses in the form of Win32 virtual key
  codes.
- displays overall and latest speed (WPM) and keystroke saving rate
  (KSR) metrics in real time
- detailed UI design and polishing is deferred to later PRs

See screenshot:
image

Fixes #125

### CODE REVIEW

**Comment by caisq** [comment]
cc @antonkast-google

File: webui/src/app/app.component.html [hunk 1]
```diff
@@ -15,6 +15,17 @@
...
+  <app-external-events-component
+      #externalEvents
+      [textEntryBeginSubject]="textEntryBeginSubject"
+      [textEntryEndSubject]="textEntryEndSubject">
+  </app-external-events-component>
+
+  <app-metrics-component
+      [textEntryBeginSubject]="textEntryBeginSubject"
+      [textEntryEndSubject]="textEntryEndSubject"
+  ></app-metrics-component>
```
Comment (antonkast-google) [inline webui/src/app/app.component.html:28]: line break

File: webui/src/app/app.component.spec.ts [hunk 2]
```diff
@@ -3,12 +3,16 @@ import {RouterTestingModule} from '@angular/router/testing';
...
+import {ExternalEventsModule} from './external/external-events.module';
+import {MetricsModule} from './metrics/metrics.module';
 
 describe('AppComponent', () => {
   beforeEach(async () => {
     await TestBed
         .configureTestingModule({
-          imports: [RouterTestingModule, AuthModule],
+          imports: [
+            RouterTestingModule, AuthModule, ExternalEventsModule, MetricsModule
```
Comment (antonkast-google) [inline webui/src/app/app.component.spec.ts:14]: Is there an ordering convention for these?

File: webui/src/app/external/external-events-component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+import {ExternalEventsComponent, getPunctuationLiteral, VIRTUAL_KEY, VKCODE_SPECIAL_KEYS} from './external-events.component';
+import {ExternalEventsModule} from './external-events.module';
+
+describe('ExternalEventsComponent', () => {
+  let textEntryBeginSubject: Subject<TextEntryBeginEvent>;
+  let textEntryEndSubject: Subject<TextEntryEndEvent>;
+  let fixture: ComponentFixture<ExternalEventsComponent>;
+  let component: ExternalEventsComponent;
+  let beginEvents: TextEntryBeginEvent[];
+  let endEvents: TextEntryEndEvent[];
```
Comment (antonkast-google) [inline webui/src/app/external/external-events-component.spec.ts:16]: Why are all these symbols scoped outside the lambda where they are used?

File: webui/src/app/external/external-events.component.html [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<style>
+
```
Comment (antonkast-google) [inline webui/src/app/external/external-events.component.html:2]: Is this file required?

File: webui/src/app/external/external-events.component.ts [hunk 5]
```diff
@@ -0,0 +1,299 @@
+/**
```
Comment (antonkast-google) [inline webui/src/app/external/external-events.component.ts:1]: Fine for now, but I am not looking forward to internationalizing this. Better would be a scheme to preserve the OS's input semantics.

File: webui/src/app/metrics/metrics.spec.ts [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+import {TextEntryBeginEvent, TextEntryEndEvent} from '../types/text-entry';
+
+import {MetricsComponent} from './metrics.component';
+import {MetricsModule} from './metrics.module';
+
+describe('MetricsComponent', () => {
+  let textEntryBeginSubject: Subject<TextEntryBeginEvent>;
+  let textEntryEndSubject: Subject<TextEntryEndEvent>;
+  let fixture: ComponentFixture<MetricsComponent>;
```
Comment (antonkast-google) [inline webui/src/app/metrics/metrics.spec.ts:13]: Same question about variable scope.

**Review by antonkast-google — APPROVED** [review]
LGTM

File: webui/src/app/external/external-events-component.spec.ts (hunk 3 above)
Comment (caisq) [reply webui/src/app/external/external-events-component.spec.ts:16]: This is a typical pattern in Jasmine in which we organize these variables that are shared between the various test methods (`it()` blocks) here. The `it()` calls and `beforeEach()` calls are just closures over them. 

This is slight different from the JUnit and PyUnit pattern, because these variables are not tied to any particular classes. But this is simply because there is no class involved here. Jasmine doesn't use classes much.

File: webui/src/app/metrics/metrics.spec.ts (hunk 6 above)
Comment (caisq) [reply webui/src/app/metrics/metrics.spec.ts:13]: See answer to the same question above.

File: webui/src/app/external/external-events.component.html (hunk 4 above)
Comment (caisq) [reply webui/src/app/external/external-events.component.html:2]: Good question. Turns out it is not required. I removed it.

File: webui/src/app/app.component.spec.ts (hunk 2 above)
Comment (caisq) [reply webui/src/app/app.component.spec.ts:14]: The ordering doesn't really matter.

Comment (caisq) [reply webui/src/app/app.component.spec.ts:14]: So I'll go with alphabetical ordering.

File: webui/src/app/app.component.html (hunk 1 above)
Comment (caisq) [reply webui/src/app/app.component.html:28]: Done.

**Comment by caisq** [comment]
Thanks for the review! Merging PR now.

File: webui/src/app/external/external-events.component.ts (hunk 5 above)
Comment (caisq) [reply webui/src/app/external/external-events.component.ts:1]: Acknowledged.
